### PR TITLE
Implement `DeepUpdate` on event and use it in `add_fields` processor

### PR DESCRIPTION
**Pull request by rdner**
## What does this PR do?

We have a couple of special cases when we set fields on the event
based on target field keys:

1. `@timestamp` must be set directly to the `Timestamp` property of
the event
2. `@metadata.*` must be put in the `Meta` map instead of `Fields`
3. The rest is put in `Fields`

Currently the `Event` struct has the `PutValue` function to account for
these special cases, however, some of the processors use
`event.Fields.DeepUpdate` which would not handle the special cases.

To make it easier to migrate those processors, the new function
`event.DeepUpdate` (that can be safely used with the
special cases) is introduced.

As an example, the `add_fields` processor is migrated to using this
new function and its tests have been updated to assert the correct behaviour.

**It's a partial solution just to demo the approach, if it's approved I will continue to migrate the rest of the processors to using the same function**

## Why is it important?

There was an issue reported earlier https://github.com/elastic/beats/issues/25425 which indicates that some customers are willing to set additional metadata with processors and not all the processors allow that. Particularly, the issue was about the `add_fields` processor.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Run filebeat with this configuration:

```yaml
filebeat:
  inputs:
    - type: log
      paths:
        - /path/to/your/file

processors:
  - add_fields:
      target: "@metadata"
      fields:
        op_type: "index"

output:
  elasticsearch:
    allow_older_versions: true
    hosts: ["localhost:9200"]
```

And run Elasticsearch in its default configuration locally.

When you write to the log file (`/path/to/your/file` in the config) before this fix you'll see a new ingested document with the field `@metadata.op_type="index"`, after applying this fix the document ingestion should fail with the following error on the console:

```
{\"type\":\"illegal_argument_exception\",\"reason\":\"only write ops with an op_type of create are allowed in data streams\"}, dropping event!"
```

which is expected because starting with v8.0.0 beats use data streams and don't allow setting `op_type` to anything but `create`. But this also means that our field we configured `@metadata.op_type="index"` was set as a part of the document metadata and not its fields which is exactly what is tested here.

## Related issues

- Relates https://github.com/elastic/beats/issues/25425

### CODE REVIEW

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -33,6 +33,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Enrich kubernetes metadata with node annotations. {pull}29605[29605]
 - Allign kubernetes configuration settings. {pull}29908[29908]
 - Remove legacy support for SSLv3. {pull}30071[30071]
+- `add_fields` processer is now able to set metadata in events {pull}30092[30092]
```
Comment (kvch) [inline CHANGELOG.next.asciidoc:36]:
```suggestion
- `add_fields` processor is now able to set metadata in events {pull}30092[30092]
```

Comment (rdner) [reply CHANGELOG.next.asciidoc:36]: 🤦 thanks!

File: CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -199,6 +200,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Add `xpack.enabled` support for Enterprise Search module. {pull}29871[29871]
 - Add gcp firestore metricset. {pull}29918[29918] 
 - Remove strict parsing on RabbitMQ module {pull}30090[30090]
+- Add gcp firestore metricset. {pull}29918[29918]
```
Comment (kvch) [inline CHANGELOG.next.asciidoc:203]: Please move this change to a separate PR. AFAIK editing multiple lines can cause issues when backporting and preparing changelogs before release.

Comment (rdner) [reply CHANGELOG.next.asciidoc:203]: Hmm, somehow rebase on master added this, weird.

File: libbeat/processors/actions/add_fields.go [hunk 3]
```diff
@@ -71,18 +71,20 @@ func NewAddFields(fields common.MapStr, shared bool, overwrite bool) processors.
 	return &addFields{fields: fields, shared: shared, overwrite: overwrite}
 }
 
-func (af *addFields) Run(event *beat.Event) (*beat.Event, error) {
+func (af *addFields) Run(event *beat.Event) (_ *beat.Event, err error) {
```
Comment (kvch) [inline libbeat/processors/actions/add_fields.go:74]: Why use named parameters? We tend to avoid them in the codebase.

Comment (rdner) [reply libbeat/processors/actions/add_fields.go:74]: that's a leftover from my first approach with the whole backup of the event. I will remove this.

File: libbeat/beat/event.go [hunk 4]
```diff
@@ -66,17 +71,88 @@ func (e *Event) GetValue(key string) (interface{}, error) {
...
-func (e *Event) PutValue(key string, v interface{}) (interface{}, error) {
-	if key == "@timestamp" {
-		switch ts := v.(type) {
-		case time.Time:
-			e.Timestamp = ts
-		case common.Time:
-			e.Timestamp = time.Time(ts)
-		default:
-			return nil, errNoTimestamp
+func (e *Event) DeepUpdate(d common.MapStr) {
```
Comment (kvch) [inline libbeat/beat/event.go:74]: Could you please add header comments to the exported functions?

**Review by kvch — COMMENTED** [review]
Minor stuff, once addressed I will approve.

**Review by kvch — APPROVED** [review]
